contracts/cw721-ics/src/contract.rs

```diff
@@ -0,0 +1,162 @@
+    }
+}
+
+fn check_can_send(
+    tract: &CW721ContractWrapper,
+    deps: Deps,
+    env: &Env,
+    info: &MessageInfo,
+    token: &TokenInfo<Empty>,
+) -> Result<(), ContractError> {
+    // Owner or minter (ICS contract) can send
+    let minter = tract.minter.load(deps.storage)?;
+    if token.owner == info.sender || info.sender == minter {
+        return Ok(());
+    }
+
+    // any non-expired token approval can send
+    if token
+        .approvals
+        .iter()
+        .any(|apr| apr.spender == info.sender && !apr.is_expired(&env.block))
+    {
+        return Ok(());
+    }
+
+    // operator can send
+    let op = tract
+        .operators
+        .may_load(deps.storage, (&token.owner, &info.sender))?;
+    match op {
+        Some(ex) => {
+            if ex.is_expired(&env.block) {
+                Err(ContractError::Unauthorized {})
+            } else {
+                Ok(())
+            }
+        }
+        None => Err(ContractError::Unauthorized {}),
+    }
+}
+
+fn transfer(
+    tract: CW721ContractWrapper,
+    deps: DepsMut,
+    env: Env,
+    info: &MessageInfo,
+    recipient: &str,
+    token_id: &str,
+) -> Result<TokenInfo<Empty>, ContractError> {
+    let mut token = tract.tokens.load(deps.storage, token_id)?;
+    check_can_send(&tract, deps.as_ref(), &env, info, &token)?;
+
+    token.owner = deps.api.addr_validate(recipient)?;
+    token.approvals = vec![];
+    tract.tokens.save(deps.storage, token_id, &token)?;
+    Ok(token)
+}
+
+fn execute_transfer(
+    tract: CW721ContractWrapper,
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    recipient: String,
+    token_id: String,
+) -> Result<Response, ContractError> {
+    // Do the transfer
+    transfer(tract, deps, env, &info, &recipient, &token_id)?;
+
+    Ok(Response::new()
+        .add_attribute("action", "transfer_nft")
+        .add_attribute("sender", info.sender)
+        .add_attribute("recipient", recipient)
+        .add_attribute("token_id", token_id))
+}
+
+fn execute_send(
+    tract: CW721ContractWrapper,
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    contract: String,
+    token_id: String,
+    msg: Binary,
+) -> Result<Response, ContractError> {
+    // Do the transfer
+    transfer(tract, deps, env, &info, &contract, &token_id)?;
+
+    // Build the message to send
+    let send = Cw721ReceiveMsg {
+        sender: info.sender.to_string(),
+        token_id: token_id.clone(),
+        msg,
+    };
+
+    // Send message
+    Ok(Response::new()
+        .add_message(send.into_cosmos_msg(contract.clone())?)
+        .add_attribute("action", "send_nft")
+        .add_attribute("sender", info.sender)
+        .add_attribute("recipient", contract)
+        .add_attribute("token_id", token_id))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    // TODO: Do we need any custom routes
```

Comment:
```suggestion
    // TODO: Do we need any custom routes
```

Not that I can think of... am I forgetting something?

Reply:
TODO: query for `CHANNELS` map.

Reply:
Also, investigate: https://github.com/public-awesome/ics721/issues/12

This is likely a pass through query to the escrow for the channel. Shouldn't need any extra state here.

Reply:
Done - see `ListChannels`, `ListClassIds`, and comment on #12 .